docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,461 @@
+# OKEP-4380: Network QoS Support
+
+* Issue: [#4380](https://github.com/ovn-org/ovn-kubernetes/issues/4380)
+
+## Problem Statement
+
+The workloads running in Kubernetes using OVN-Kubernetes as a networking backend might have different requirements in handling network traffic. For example video streaming application needs low latency and jitter whereas storage application can tolerate with packet loss. Hence NetworkQoS is essential in meeting these SLAs to provide better service quality.
+
+The workload taffic can be either east west (pod to pod traffic) or north south traffic (pod to external traffic) types in a Kubernetes cluster which is limited by finite bandwidth. So NetworkQoS must ensure high priority applications get the necessary NetworkQoS marking so that it can prevent network conjestion.
+
+## Goals
+
+- Provide a mechanism for users to set DSCP marking on east west and north south traffic.
+- Provide a mechanism for users to set Metering on east west and north south traffic.
+- Make NetworkQoS CRD fields to be more extensible.
+
+## Non-Goals
+
+- Ingress Network QoS.
+
+- Consolidating with current `kubernetes.io/egress-bandwidth` and `kubernetes.io/ingress-bandwidth` annotations.
+Nonetheless, the work done here does not interfere with the current bandwidth NetworkQoS mechanism.
+
+- The DSCP marking does not need to be handled or acted upon by OVN-Kubernetes, just added to selected headers.
+
+## Introduction
+
+To enable NetworkQoS, we will use Differentiated Services Code Point (DSCP) which allows us to classify packets by setting a 6-bit field in the IP header, effectively marking the priority of a given packet relative to other packets as "Critical", "High Priority", "Best Effort" and so on.
+
+## User-Stories/Use-Cases
```

Comment:
ok, now added only two user stories with intent to be crisp and clear, is that ok now ?

Reply:
I would still like to understand WHY we want to use DSCP, who will use it and how?

Reply:
@npinaeva we use it heavily in our data centers to differentiate the traffic between storage and stream. 
For true end-to-end differentiation, we need to mark the packet on the chassis where the pod is running

```
                                             .---------------.      
+---------+                              _.-'                 `--.  
|       +-+---+    +------+   +----+    /         overlay         \ 
|   Pod |eth0 +----+br-int+---|NIC |--->       over network        )
|       +-+---+    +------+   +----+    `.        fabric         ,' 
+---------+                               `--.               _.-'   
                                              `-------------'       
```
The DSCP marking through the `inherit` setting of OVS is copied over to the outer packet.  The NICs act on the marking and enforce traffic policing. The same packet is then policied in the underlay by the network switches.

Furthermore, one could write applications, for example a gateway, that could treat packets differently based on these marks. 

So, it is not just for QoS, but there are other uses cases of treating the packet's forwarding behaviour using iptables on the gateway.

Reply:
Also, I recently watched a really good presentation on this use case:  https://youtu.be/76tfwAvDWX8?feature=shared   :)

Reply:
FYI: I added a section on doc to reflect this topic.

Reply:
Thank you for this example! I think we are missing an Introduction section https://github.com/ovn-kubernetes/ovn-kubernetes/blob/master/docs/okeps/okep-4368-template.md?plain=1#L25 that should clarify any details needed to understand the context. 
I think your use case should go in the Introduction, and besides just saying that is how "we" use DSCP, it would be nice to explain why users of ovn-k in general may want to use DSCP (it is a well-known mark for QoS in the networking world, so I am sure this part won't be a problem, but it should be clearly stated, so if someone asks "why did you choose DSCP and not something else" this section should answer that. For now DSCP just appears without any introduction.